### DISCUSSION TITLE
Protecting from catastrophic failure when goal=1?

### DISCUSSION THREAD

**Discussion by OkJaybird** [moosefs/moosefs#404] [Q&A]
I am loving moosefs and have a unique scenario I'm curious to get advice on.

I have a large collection of report files, each roughly ~50GB in size. The nature of the data is reproducable, so I've set the goal for these files to 1 in order to save space. I'm not too concerned with having to regenerate a handful of the files in the event of disk failure / data loss, but losing all of them simultaneously would not be ideal, since they take a while to generate and the computation puts additional load on our systems.

I can imagine a scenario where a disk containing chunks for all the data files fails, and thus all the data files are corrupted simultaneously. Does moosefs have any mechanisms to protect from a scenario like this?  Maybe a way to restrict the number of disks a given file is spread across? From the documentation, it looks like increasing the chunkerver's `HDD_RR_CHUNK_COUNT` parameter may help encourage moosefs to consolidate a file's chunks across fewer disks. I'm wondering if there are additional options to help limit the blast radius for these types of files when data loss does occur.

**Comment by chogata** [comment]
You would basically need a mechanism that ensures, that if the first chunk of file X is stored on chunkserver A, and the first chunk of file Y is stored on chunkserver B, then the second chunk of file X is also stored on chunkserver A and on the same hard drive/mountpoint (the same path from `hdd.cfg`) and the second chunk of file Y is stored on chunkserver B on the same hard drive/mountpoint. And so on with each next chunk of files X and Y. Then, when one of the drives fails, all chunks of either file X or file Y are missing, not one chunk from each, meaning you only have to re-create either X or Y.

The problem is, that currently MooseFS doesn't "care" if a newly created chunk is a next chunk of a file that is already stored somewhere. We got this request before and we were thinking about the possibility of adding such a mechanism, however, there is the problem of replications (both recovery replications for undergoal chunks and rebalancing replications, to keep the space usage even among chunkservers and drives) and the problem of chunks that belong to more than one file (common when you are using snapshots). Basically there is no good solution. We were thinking about adding a mechanism that would ensure the rule "keep one file on one disk" during the create operation, but the users would have to take into account that this would kind of "deteriorate" over time, due to replications.

And no, changing `HDD_RR_CHUNK_COUNT` won't do anything in this case. This variable only says how many chunk files will be recorded in one directory on a certain mountpoint (one of the `00` to `FF` directories that chunkserver keeps the chunks in) before moving on to another directory (the one with smallest amount of chunks at this point in time). But this happens only after the master has chosen the chunkserver(s) to record the new chunk on and the chunkserver has chosen the mountpoint (hard drive) to record the chunk on.

**Comment by chogata** [comment]
The one thing you could do would be to label each chunkserver with a different label, create simple storage classes for them (keep one copy on a certain label, one class for each label you have) and manually "distribute" the report files by assigning each of them a different class, spreading them evenly between chunk servers. This would ensure all chunks of a single report will be kept on the same chunk server, as this will be the only one with this label, but if the chunk server has more than one mountpoint (hard drive), you have no control over which chunks will be stored on which drives.

**Comment by dagelf** [comment]
Or just keep those files on a separate drive. Or run mergerfs over moosefs so you can put your 1-copy-tree on a certain drive. Or just stick to goal 2 minimum for everything. Although I would not advise anything less than goal 3. Remember your time is more valuable than the cost of hard drives... if regenerating those reports will take your time, you should think carefully.... even just sorting data and making things accessible takes valuable time.

**Comment by OE2WNL** [comment]
You could also set Label in the config of your chunkservers (A,B,C) and then create the directories FolderA,FolderB,FolderC and set `mfssetsclass 1A FolderA`
for the FolderA directory and 1B for the FolderB directory ...
So files within the FolderA will land on chunkserver with Label A and so on.
might this help?

For the sake of simplicity I would define Goal=2 as minimum so you don't have to think about it anymore but this might be a solution if you stick to your 1 Goal strategy

Greetings
-Wolfgang
